Add 16S processing overview tutorial

cduvallet opened the pull request:
Bod and I wrote this broad overview/tutorial, which covers QIIME 2 functionalities through the lens of a typical 16S workflow. It has a lot of overlap with the new [overview tutorial](https://docs.qiime2.org/2018.6/tutorials/overview/), but is slightly more catered to experienced microbiome users who are considering switching to QIIME 2.

Let us know the best way to move forward on this - would be great to have this included in the next release, but if there are edits to think about/make, we're happy to post this on the forum while we figure out the best way to process this.

cduvallet commented:
Hi @nbokulich - thanks so much for the detailed and thoughtful comments! I haven't looked through all the in-line comments but I do agree with your general comments up top. I like the idea of having two tutorials: one for beginners and one for experts/"people who don't want emojis in their tutorials 😂". 

I agree that adding in more specific info on *why* experts should switch to QIIME 2 in the prelude will be a good addition, and that a lot of the extra details can be moved to the beginniner-targeted overview tutorial.

I'm happy to take the lead on making those edits, though I am going to be away next week so probably won't have too much time to get to it (perhaps this might need to wait until the next QIIME 2 release, then?) Logistically speaking, I should just make those edits and keep pushing to this branch for the changes to show up here, right?

Also, I may add some section headings to the original overview tutorial in the course of these edits. It'll be helpful to me as I link to different sections of the tutorial, and will also hopefully make the overview slightly less wall-of-text and easier to parse. Does that sound ok to y'all?

nbokulich commented:
> I agree that adding in more specific info on why experts should switch to QIIME 2 in the prelude will be a good addition

See or link to [here](https://docs.qiime2.org/2018.6/about/) and [here](https://docs.qiime2.org/2018.6/concepts/) as that covers some of the strengths of QIIME 2 (probably worth reiterating whatever you see as the main strengths for expert users, and link there for some more details). Also mention the [multiple interfaces](https://docs.qiime2.org/2018.6/interfaces/). [Semantic types](https://docs.qiime2.org/2018.6/semantic-types/) could also be a strength, though maybe not to expert users.

> (perhaps this might need to wait until the next QIIME 2 release, then?)

Alas it may 😢 — this can be made into a "community tutorial" on the forum if you want it live faster. But let's see — since I've given this a first review it should be faster from here and we may be able to squeak this under the door and into this release.

> Logistically speaking, I should just make those edits and keep pushing to this branch for the changes to show up here, right?

Yes!

> Also, I may add some section headings to the original overview tutorial in the course of these edits. It'll be helpful to me as I link to different sections of the tutorial, and will also hopefully make the overview slightly less wall-of-text and easier to parse. Does that sound ok to y'all?

❤️❤️❤️ Absolutely... the overview tutorial has a long preamble (pre-ramble?) of sort of warning and caveat-type info — much of that could/should be parsed by headings and/or moved around. These tutorials are by no means untouchable — please do fix that tutorial anywhere you see fit!

Thanks!

mestaki commented:
Thanks @nbokulich! We were mid writing this tutorial when your overview tutorial came out 🤷‍♂️  thus the abundance of overlap between the 2. At the time we both agreed that there were important information still in there that could be useful for more advanced users, which is what we aimed for but of course the line between beginners and experts can be murky. Especially with new fundamental concepts like semantic types, provenance, ASVs, and plugins which would be new even for experts. I'm honestly happy to move forward with any of the proposed options. I agree with Claire though that it would be useful to have a separate advanced tutorial, for example, remember the moment of nirvana she experienced in SD when she realized qza were just zipped files? I imagine that kind of info is very pertinent to advanced users and perhaps don't need to be hidden in a pile of other basic concepts in a tutorial. But we can also totally do both and add a lot of this material to the current overview tutorial as well as fine-tune an 'advanced' one for users who want to skip familiar concept and just go to the new ones. 
That all being said, I'd love to help in any way but I've now just switched gears to thesis-writing mode with a rather tight deadline so I'm not sure how much free time I'll be able to dedicate here until that's done (~last week of September).

nbokulich commented:
@mestaki I DO really like the idea of an expert vs. beginner tutorial and I DO really like this tutorial — as I said above I am just concerned about the large degree of overlap between the two and lack of clear distinction for different user types.

So I think the "middle way" I proposed above is best — merge some parts of this tutorial with the overview tutorial to trim down this tutorial and make it more geared toward experts (and many parts of your tutorial, like importing and read joining, are not covered very well in the overview tutorial but are relevant to both). But I am happy with any of the three options:
1. just address the review comments and move on (we can consider migrating text in a future release)
2. ~merge with the overview 100%~ (we can all disagree on this one — we all like the beginner/expert distinction)
3. Move some text into overview, focus on expert users in this tutorial

It could also be worth moving that text to the specific "technical" tutorials, e.g., the read joining tutorial, as some of the technical tutorials are sort of light on the theory and discussion. The "beginner" (current overview) and "expert" (this) tutorials could then both just link to the technical tutorial. But that is a more ambitious proposal — we can start by building the beginner/expert tutorials and port text over in a future release (because the overview tutorial is already quite long and verbose in parts)

cduvallet commented:
Sorry for being MIA - had much less access to internet on my trip than I expected!

I'll start with the first option (addressing the review comments) and see where things stand after that. I think that moving some of this text into the overview tutorial is also very doable (and I'd be excited to help improve that tutorial as well) - will just be a matter of bandwidth more than anything.

tl'dr - it's on my to-do list and I will ping y'all again when there have been updates!

When is the next QIIME release planned for? October?

nbokulich commented:
Thanks @cduvallet !

Yes — next release is planned for the end of October. Current schedule:
Submit PRs by: 10/17
Merge PRs by: 10/22
Release Day: 10/24

cduvallet commented:
Alright, just took a first pass at writing the "expert" tutorial.

Quick question: how can I preview these changes locally? I'm not as familiar with rst as markdown, so have probably messed up a lot of the formatting, but am not sure how to preview the compiled doc on my own computer.

nbokulich commented:
to build a preview, see the "contributing" section [here](https://github.com/qiime2/docs)... ignore the publishing section. Thanks! I will take a look Monday.

thermokarst commented:
Yikes, those README instructions are a bit out of date. Start with a q2 env, activate, then install the requirements.txt jazz, then you can run the make commands.

nbokulich commented:
Hey @cduvallet it looks like `make preview` fails because you use `:docs:` instead of `:doc:` in a few places. Let me know when `make preview` is passing and I will start my review.

If you have any command blocks in the tutorial you can use `make preview DEBUG=tutorials/overview` to test the commands.

Thanks!

cduvallet commented:
Ahh, I see! Okay, fixed the bugs so that `make preview` works now.

There are still some outstanding to do's but I think the bulk of it is ready for review.

I'm also definitely not linking to the plugins correctly. Looks like the right way might be `:doc:'link text <../plugins/available/plugin_name/function_name/>'` - is that right? I'm guessing the list of plugins gets generated somewhere else since there's nothing in the plugin folder.

nbokulich commented:
Awesome! I will begin review — should I wait until you fix the plugin links?

You have the right format for linking to plugins, except the single quotes should be backticks. E.g., ``:doc:`collapse taxonomy <../plugins/available/taxa/collapse/>` ``

The plugin API docs are automatically generated, so are not physically stored in the repo.

cduvallet commented:
Ok, almost done updating the plugin links but I'm getting an error when I try to link to the entire plugin rather than a method.

e.g. ``:doc:`q2-demux <../plugins/available/demux/emp-paired/>` `` works, but ``:doc:`q2-demux <../plugins/available/demux/>` `` does not. It gives the following error during `make preview`:

```
WARNING: unknown document: ../plugins/available/demux/
```

nbokulich commented:
Hey @cduvallet ,
Courtesy of @thermokarst : you need to add an `index` on the end of that filepath. E.g.,
``:doc:`q2-demux <../plugins/available/demux/index>` ``

cduvallet commented:
Sweet, thanks! (Making a mental note to include some write-up of all the things I've learned about linking in rst docs to the qiime2/docs README)

Links are all updated and the doc should be ready for review. Note that there are many places that I say "this is discussed more in the overview tutorial" but that aren't linked anywhere. I haven't gone through and ported the text I removed to the generic overview tuorial yet, that might take me a while bc it will take more effort than this! :P

I'm also considering adding a tl'dr somewhere (beginning? end?) with just the list of steps and respective plugins. Thoughts on this?

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+   -  quality filtering, length trimming, and clustering with VSEARCH or dbOTU
+
+5. Assigning taxonomy
+
```
maybe mention at the end: "See the ``:doc:`overview tutorial <blah>` `` and ``:doc:`available plugins <blah>` `` for additional optional downstream analysis steps."

You cover the important ones here, but just want to make sure users reading this don't get the impression that that's all.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+This step has the potential to be the most confusing part of the QIIME 2 pipeline as there are dozens of import and format types to choose from.
+To see a full list of available import/format types use: ``qiime tools import --show-importable-formats`` and ``qiime tools import --show-importable-types``
+
+If you're importing data that you've generated, you'll likely need to generate a :ref:`manifest file <manifest file>`, which is just a text file that  maps each FASTQ or FASTA file to its sample ID and direction (if applicable).
```
If you're importing data -> If you're importing FASTQ data

nbokulich replied:
manifest files are only for fastq I believe (or so the format name would imply). @ebolyen @thermokarst is there an exception I'm not aware of?

If so, change:
that  maps each FASTQ or FASTA file to its sample ID -> that maps each FASTQ file to its sample ID

And maybe add a line about importing demultiplexed fasta data (can just point to the OTU picking tutorial, which describes that process)

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+The processing steps we'll cover in this tutorial include:
+
+1. Importing data into QIIME 2
```
Importing raw sequence (FASTQ) data into QIIME 2

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+You can learn more about how ``cutadapt`` works under the hood by reading their `documentation <https://cutadapt.readthedocs.io/en/stable/index.html>`__.
+
+Note: Currently ``q2-demux`` and ``q2-cutadapt`` do not support demultiplexing dual-barcoded paired-end sequences, but only can demultiplex with barcodes in the forward reads.
+So for the time being for this type of demultiplexing needs to be done outside of QIIME 2 using other tools, for example
```
So for the time being for this type of demultiplexing needs to be done -> So for the time being this type of demultiplexing needs to be done

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+Merging reads
+~~~~~~~~~~~~~~
+
+Whether or not you need to merge reads depends on how you plan to cluster or denoise your sequences into amplicon sequence variants (ASVs) or operational taxonomic units (OTUs).
```
Add this? (or blend with the line below)
"If you plan to use deblur or OTU clustering methods next, join your sequences now. If you plan to use dada2 to denoise your sequences, _do not merge_ — dada2 performs read merging automatically after denoising each sequence.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+The :doc:`q2-cutadapt <../plugins/available/cutadapt/index>` plugin has comprehensive methods for removing non-biological sequences from :doc:`paired-end <../plugins/available/cutadapt/trim-paired/>` or :doc:`single-end <../plugins/available/cutadapt/trim-single/>` data.
+
+If you're going to use DADA2 to denoise your sequences, you can remove biological sequences at the same time as you call the denoising function.
+All of DADA2's ``denoise`` fuctions have some sort of ``--p-trim`` parameter you can specify to remove base pairs from the end(s) of your reads.
```
trims 5' end only

(you could mention truncate, but maybe best to mention with the note below)

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+If you're going to use DADA2 to denoise your sequences, you can remove biological sequences at the same time as you call the denoising function.
+All of DADA2's ``denoise`` fuctions have some sort of ``--p-trim`` parameter you can specify to remove base pairs from the end(s) of your reads.
+(Deblur does not have this functionality yet.)
```
Add this?
"Note that dada2's trim parameter only trims the 5' end. Although the `--p-truncate` parameter can be used similarly to trim the 3' ends of sequences, read length heterogeneity can make it very difficult to predict the appropriate truncation site for removing non-biological sequence data. In this situation, it may be preferable to use q2-cutadapt to trim non-biological sequence data prior to using dada2."

cduvallet replied:
Ok, I added the clarification that `--p-trim-left` only trims the 5' end.

But I'm confused about the `--p-truncuate` statement here - why should people use cutadapt to truncate reads rather than passing in an appropriate parameter to `--p-truncate`? Seems to me like the workflow that makes the most sense is to keep things simple and do things all within dada2...?

ebolyen replied:
Yep DADA2 is usually enough to handle both cases.

cutadapt is mostly for the case where your reads will run through the reverse-complemented opposing-primer (common with ITS). It isn't something you'd worry about with 515-806, but other primer sets can have the problem.

cduvallet replied:
ok, i won't add any mention of the `--p-truncate` parameter here because we discuss it below in the dada2 part :)

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are two main approaches for grouping similar sequences together: denoising and clustering.
+The overview tutorial provides more in-depth discussion of these approaches.
```
maybe link to the specific section in the overview tutorial with a :ref:

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+~~~~~~~~~
+
+DADA2 and deblur are currently the two denoising methods available in QIIME 2.
+You can read more about the methods in the overview tutorial.
```
link to section?

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+Denoising requires little data preparation.
+Both DADA2 and deblur perform quality filtering, denoising, and chimera removal, so you shouldn't need to perform any quality screening prior to running them.
+That said, the official QIIME 2 tutorial recommends doing an initial :ref:`quality-filter <moving pictures deblur>` with default settings prior to using deblur (as recommended by the deblur developers).
```
This _is_ an official tutorial now! Maybe just say that the deblur developers recommend this.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+Denoising requires little data preparation.
+Both DADA2 and deblur perform quality filtering, denoising, and chimera removal, so you shouldn't need to perform any quality screening prior to running them.
+That said, the official QIIME 2 tutorial recommends doing an initial :ref:`quality-filter <moving pictures deblur>` with default settings prior to using deblur (as recommended by the deblur developers).
+In our experience, DADA2 performs better without this step.
```
dada2 actually has this type of QC (or at least Q-score based filtering; with optional Q-score based trimming) built in. So just state that this is entirely unnecessary prior to dada2.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+That said, the official QIIME 2 tutorial recommends doing an initial :ref:`quality-filter <moving pictures deblur>` with default settings prior to using deblur (as recommended by the deblur developers).
+In our experience, DADA2 performs better without this step.
+
+Both methods have an option to truncate your reads to a constant length (which occurs prior to denoising).
```
they have different names (truncation for dada2, trim for deblur) so maybe give the parameter names here.

cduvallet replied:
I'm adding this sentence, these are the right parameters, yeah?

> In DADA2, this is the `--p-trunc-len` parameter; in deblur it's `--p-trim-length`.

nbokulich replied:
👍

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+In our experience, DADA2 performs better without this step.
+
+Both methods have an option to truncate your reads to a constant length (which occurs prior to denoising).
+The truncating parameter is required for deblur and optional for DADA2.
```
that parameter can be disabled for deblur, so you can probably just drop this line.

nbokulich replied:
or at least that's what the docs suggest... I vaguely recall there being some mention on the forum that even though you can switch this off it leads to other issues with deblur:
```
  --p-trim-length INTEGER         Sequence trim length, specify -1 to disable
                                  trimming.  [required]
```

cduvallet replied:
mmk, if I add this sentence do you think that addresses the concerns?

> The truncating parameter is optional for both DADA2 and deblur (though if you're using deblur you'll need to specify `--p-trim-length -1` to disable truncation).

nbokulich replied:
👍

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+deblur can currently only denoise single-end reads.
+It will accept unmerged paired-end reads as input, it just won't do anything with the reverse reads.
+Note that deblur *can* take in *merged* reads and treat them as single-end reads.
```
"so merge paired-end data prior to denoising with deblur."

is that kicking the horse?

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+We discussed merging paired-end reads and removing non-biological sequences above (Sections `Merge reads`_ and `Remove non-biological sequences`_).
+
+Once your data is ready, you'll also need to dereplicate your reads before clustering.
```
data is -> data are

cduvallet replied:
lol

cduvallet replied:
http://www.smbc-comics.com/comic/they

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+'''''''''''''''''
+
+You can perform different types of quality filtering with the :doc:`quality filter <../plugins/available/quality-filter/index>` plugin.
+**TODO: what's the difference between the q-score and q-score-joined? Why are there two separate functions?**
```
@wasade could you please answer this?

wasade replied:
One is for the single end semantic type, the other is for paired IIRC

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+Sequences can be clustered *de novo* based on their genetic similarity alone (i.e. with VSEARCH) or based on a combination of their genetic similarity and abundance distributions (i.e. with distribution-based clustering).
+
+-  **Similarity-based clustering.** The QIIME 2 VSEARCH plugin method :doc:`cluster-features-de-novo <../plugins/available/vsearch/cluster-features-de-novo/>` clusters OTUs. You can change the genetic similarity threshold with the ``--p-perc-identity`` parameter. The plugin wraps the VSEARCH ``--cluster_size`` function.
```
QIIME 2 VSEARCH -> q2-vsearch

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+It will accept unmerged paired-end reads as input, it just won't do anything with the reverse reads.
+Note that deblur *can* take in *merged* reads and treat them as single-end reads.
+
+OTU Clustering
```
link to the OTU clustering tutorial somewhere in here?

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+This method wraps the ``--usearch_global`` VSEARCH function.
+You can decide which reference database to cluster against with the ``--i-reference-sequences`` flag.
+The input file to this flag should be a ``.qza`` file containing a fasta file with the sequences to use as references, with QIIME 2 data type ``FeatureData[Sequence]``.
+Most people use GreenGenes or SILVA, but others curate their own databases or use other standard references (e.g. UNITE for ITS data).
```
Most people use GreenGenes or SILVA -> Most people use GreenGenes or SILVA for 16S rRNA gene sequences

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+- :doc:`classify-consensus-blast <../plugins/available/feature-classifier/classify-consensus-blast/>`: BLAST+ local alignment
+- :doc:`classify-consensus-vsearch <../plugins/available/feature-classifier/classify-consensus-vsearch/>`:  VSEARCH global alignment
+
+Both use the *consensus* approach of taxonomy assignment, which you can learn more about in the overview and tweak with the ``maxaccepts``, ``perc-identity``, and ``min-consensus`` parameters.
```
link to overview section?

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+These two functions differ in the type of machine learning model that they use.
+(To get more details on the specific models implemented, check out `the plugin's associated paper <https://doi.org/10.1186/s40168-018-0470-z>`__.)
+These methods need a pre-trained model as one of the inputs: you can either download one of the pre-trained taxonomy classifiers from the :doc:`data resources page <../data-resources/>`, or train one yourself (following the steps outlined in the :doc:`feature classifier tutorial <feature-classifier>`).
```
those methods fit the models... then ``classify-sklearn`` is used to classify with any fitted model.

So maybe just point to the ``classify-sklearn`` method above instead of modifying the text here (sure, ``fit-classifier-sklearn`` is there for anyone to use but requires some undocumented parameter specifications to really have any fun)

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+(To get more details on the specific models implemented, check out `the plugin's associated paper <https://doi.org/10.1186/s40168-018-0470-z>`__.)
+These methods need a pre-trained model as one of the inputs: you can either download one of the pre-trained taxonomy classifiers from the :doc:`data resources page <../data-resources/>`, or train one yourself (following the steps outlined in the :doc:`feature classifier tutorial <feature-classifier>`).
+
+Analyze feature table and gain insight
```
insight is just begging for an emoji.

nbokulich replied:
:godmode:

cduvallet replied:
i'll give you data are if you give me emoji-free tutorial ;)

nbokulich replied:
😥

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+Analyze feature table and gain insight
+--------------------------------------
+
+At this point, you should be ready to analyze your feature table to answer your scientific questions!
```
🎉

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+--------------------------------------
+
+At this point, you should be ready to analyze your feature table to answer your scientific questions!
+QIIME 2 offers multiple built-in functions to analyze your data, and you can also `export <Export the data>`_ it to do downstream analyses in your preferred coding language.
```
coding language / software package.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+Some general things you can do with QIIME 2 are:
+
+-  **Look at the data:** QIIME 2 has some a nice :doc:`taxa barplot visualizer <../plugins/available/taxa/barplot/>` to make visually exploring your data easy. You can also visualize your data on a PCoA plot with the :doc:`emperor <../plugins/available/emperor/plot/>` plugin (after calculating beta diversity between samples).
```
has some a nice -> has a nice

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+
+-  **Look at the data:** QIIME 2 has some a nice :doc:`taxa barplot visualizer <../plugins/available/taxa/barplot/>` to make visually exploring your data easy. You can also visualize your data on a PCoA plot with the :doc:`emperor <../plugins/available/emperor/plot/>` plugin (after calculating beta diversity between samples).
+-  **Build a phylogenetic tree:** QIIME 2 has a :doc:`phylogeny <../plugins/available/phylogeny/index>` plugin with different tree-building methods.
+-  **Calculate alpha diversity of your samples:** the :doc:`diversity plugin <../plugins/available/diversity/index>` has many `alpha diversity metrics <https://forum.qiime2.org/t/alpha-and-beta-diversity-explanations-and-commands/2282>`__ available through the ``alpha`` and ``alpha-phylogenetic`` methods.
```
the `core-metrics*` pipelines are somewhat more user friendly and streamlined (since e.g., beta is followed by pcoa is followed by emperor). Could link to the flowchart in the overview tutorial if it's useful here.

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+-  **Calculate alpha diversity of your samples:** the :doc:`diversity plugin <../plugins/available/diversity/index>` has many `alpha diversity metrics <https://forum.qiime2.org/t/alpha-and-beta-diversity-explanations-and-commands/2282>`__ available through the ``alpha`` and ``alpha-phylogenetic`` methods.
+-  **Calculate beta diversity between samples:** the :doc:`diversity plugin <../plugins/available/diversity/index>` also has these metrics available in the ``beta``, ``beta-phylogenetic``, and ``beta-phylogenetic-alt`` methods.
+-  **Test for differences between samples**, through differential abundance or distribution testing: PERMANOVA, ANOSIM, ANCOM, and Gneiss are some of the relevant methods which are available in QIIME 2. PERMANOVA and ANOSIM can be done with the :doc:`beta-group-significance <../plugins/available/diversity/beta-group-significance/>` method in the ``diversity`` plugin. ANCOM is available in the :do ...
+-  **Build machine learning classifiers to make predictions:** the :doc:`q2-sample-classifier <../plugins/available/sample-classifier/index>` plugin has several actions for these classifiers, and the associated :doc:`"Predicting sample metadata values with q2-sample-classifier" tutorial <sample-classifier>` provides more details.
```
classifiers and regressors

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+While ``export`` only outputs the data, the :ref:`extract <export vs extract>` tool allows you to also extract other metadata such as the citations, provenance etc.
+
+Note that this places generically named files (e.g. ``feature-table.txt``) into the output directory, so you may want to immediately rename the files to something more information (or somehow ensure that they stay in their original directory)!
+(**TODO: does the new 2018.8 version let you rename files?**)
```
@turanoo or @thermokarst do you want to answer this?

nbokulich commented on source/concepts.rst:
```diff
@@ -34,6 +34,8 @@ Semantic types also help users avoid semantically incorrect analyses. For exampl
 
 Check out the :doc:`semantic types page <semantic-types>` for more information about semantic types and what types are currently available.
 
+.. _`plugins`:
```
I don't see a :ref: for this below... is this used?

cduvallet replied:
nope! i wrote this before I knew how to link to the list of plugins...

nbokulich commented:
>I haven't gone through and ported the text I removed to the generic overview tuorial yet, that might take me a while bc it will take more effort than this! :P

That's fine... not a requirement for this doc at all. I am looking forward to those changes too 😄 . Let me know if I can help.

>I'm also considering adding a tl'dr somewhere (beginning? end?) with just the list of steps and respective plugins. Thoughts on this?

I think the TOC does an admirable job of this... you could manually build a TOC so that you could include additional content (e.g., sublists with names of the plugins/actions used) but I don't think it's necessary. Having TOC + tl'dr could be a case of excessive front matter.

cduvallet commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+QIIME 2 also encourages making **microbiome research more reproducible**.
+QIIME 2 reduces inappropriate analyses by defining specific data types and restricting methods only to their appropriate data type inputs.
+It also keeps track of everything that's been done to a given data file in the :doc:`provenance <../concepts>` metadata associated with each QIIME 2 artifact.
+Furthermore, by wrapping tools into a common framework, **data processing pipelines are streamlined**: with QIIME 2, you'll be able to re-write what was once a collection of scripts in different coding languages, calling different executables and re-formatting data inputs and outputs in most intermediate steps, into one (or a few) simple bash scripts.
```
hm yes, I see how a skeptic would respond. This is still a HUGE drawing point (at least for me), so I wanna emphasize it a bit. How about this:

> Furthermore, by wrapping tools into a common framework, **data processing pipelines are streamlined**. With QIIME 2, most data processing workflows can be consolidated into one (or a few) bash scripts, reducing the number of different programs or executables you need to call and re-formatting data inputs and outputs that are necessary.

ebolyen commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -0,0 +1,296 @@
+If you need to process your data more interactively, you might want to use the Python API - it is much faster since objects can be simply stored in memory.
+You can learn more about the different `QIIME 2 interfaces <https://docs.qiime2.org/2018.6/interfaces/>`__.
+
+**Pro-tip #3: to enable tab-complete in QIIME 2**, run ``source tab-qiime``.
```
yep, that is the case (sorry I am behind on every notification there is :persevere:)

cduvallet commented:
Ok, made most of those changes (except a few unresolved comments, still open above).

I also added a list of "relevant plugins" under each section header. I noticed there is no doc page for `qiime tools`?! Would it be possible to generate this without it being a huge hassle...? I'd love to be consistent and not have to link people to the importing/exporting tutorials...

nbokulich commented on source/tutorials/qiime2-for-experienced-microbiome-researchers.rst:
```diff
@@ -204,41 +232,54 @@ There isn't currently a QIIME 2 function to trim reads to the same length withou
+
 You can perform different types of quality filtering with the :doc:`quality filter <../plugins/available/quality-filter/index>` plugin.
-**TODO: what's the difference between the q-score and q-score-joined? Why are there two separate functions?**
+The `q-score` method is for single- or paired-end sequences (i.e. `SampleData[PairedEndSequencesWithQuality | SequencesWithQuality]`) while `q-score-joined` is for merged reads (i.e. `SampleData[JoinedSequencesWithQuality]`).
```
is for merged reads --> is for joined reads

sorry to get pedantic... both terms are used, but we prefer using "joined" paired-end reads to avoid confusion with "merging" e.g., feature tables or sequence files.

ebolyen commented:
@nbokulich, the tools are described nowhere actually. We should fix that! It would probably make sense here: https://docs.qiime2.org/2018.8/interfaces/q2cli/ and we could do the same help-text generation for the most part I expect. We really don't have a lot of documentation on any of the interfaces which is a shame.

cduvallet commented:
Ok, looks like the only thing left is to confirm whether you can rename files when exporting. I just tried it, and it looks like the answer is no (you can specify a directory, but the files exported into that directory are still named something generic like `feature_table.biom`). @thermokarst let me know if that's not quite the right interpretation!

@ebolyen - getting docs made for the qiime tools would be really great. This was one of my *biggest* frustrations when I first tried to start using qiime2 (and is why I gave up almost immediately the first time) - the first obstacles I encountered happened during import, and I couldn't find the documentation I expected.

ebolyen commented:
Re: renaming files, I think that only works if you set the `output-format` to be a FileFormat rather than leaving it default which will be a DirectoryFormat of some kind. If your output-format is a file format, then whatever path you provide should be the name it gets.

It's a bit obtuse, but maybe a solution would be to automatically "downcast" a SingleFileDirectoryFormat into the corresponding file-format when exporting. I *think* that's pretty easy to do.

cduvallet commented:
> I think we are ready to merge! Just want to confirm @cduvallet and @ebolyen that you are both ready

Hurray! Yep, ready on my end! 😄

ebolyen commented:
Ditto!

On Tue, Oct 23, 2018, 6:31 PM cduvallet <notifications@github.com> wrote:

> I think we are ready to merge! Just want to confirm @cduvallet
> <https://github.com/cduvallet> and @ebolyen <https://github.com/ebolyen>
> that you are both ready
>
> Hurray! Yep, ready on my end! 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qiime2/docs/pull/345#issuecomment-432479727>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADyuZDLaJw9ON82ztlUXzh2GlkGdNMS8ks5un8L6gaJpZM4WGxtd>
> .
>

cduvallet commented:
Ack oh no, just noticed these two scraps that should be removed from the final tutorial:

```
Importing data into QIIME 2
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Relevant plugin**: ``qiime tools import`` (@Team - there is no doc page for these??!)
```

and 

```
Export the data
~~~~~~~~~~~~~~~

**Relevant plugin**: ``qiime tools export`` (this one also doesn't have a doc page?!)
```

My bad for not catching this before the merge!! 🙈

ebolyen commented:
Oh well, we can patch the HTML (and by we, I mean @thermokarst :wink:) on the live site.
@cduvallet, would you be able to write a short PR to patch the source? Thanks!
